flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -666,10 +671,11 @@ else if (!props.forceCheckpoint()) {
 	 * Receives a {@link DeclineCheckpoint} message for a pending checkpoint.
 	 *
 	 * @param message Checkpoint decline from the task manager
+	 * @return <code>true</code> if should fail the job
 	 */
-	public void receiveDeclineMessage(DeclineCheckpoint message) {
+	public boolean receiveDeclineMessage(DeclineCheckpoint message) {
```

Comment:
This level of abstraction feels strange to me, that the return value from the decline method of the checkpoint coordinator should decide of the job fails. Conceptually, shouldn't that rather be an optional reaction of the job master to a failed `CheckpointExecuteResult`?

Reply:
@StefanRRichter You are right. The expected thing you mentioned is our second step, we will introduce a `CheckpointFailureManager` that will decide how to process failure. Currently, based on @azagrebin 's suggestion we should refactor the failure handling to prepare for the second step, the design document is here : https://docs.google.com/document/d/1ce7RtecuTxcVUJlnU44hzcO2Dwq9g4Oyd8_biy94hJc/edit?usp=sharing

Reply:
Alright, I had another look into the design doc and the doc is talking about introducing an checkpoint failure manager in a second step. So my assumption would be that this PR is mainly about introducing the unified `PendingCheckpoint#abort` and `CheckpointAbortReason`. Then, as a second step, I would expect the introduction of of the failure manager. And only as a third step I would expect the code wiring all together and changing the behaviour. Otherwise, as it is done now, we already have the changed behaviour but not all refactorings completely in place. But in order to get the behaviour changed you introduced an intermediate state of flawed abstraction that can still linger in the code for a while. Can we separate this as I suggested and only complete the work's behavioural change at the very end, when also the manager is in place?

Reply:
Maybe shorten this PR by reverting the behaviour changes, then open a PR that introduces the manager, then add a third PR redoing the behaviour changes. Like that I also don't need to imagine how the code is intended to look at some point in the future, but can comment what is in front of me :)

Reply:
@yanghua I think @StefanRRichter is right about the steps order. The confusion probably comes from the fact that section [T1-5] in design docs should actually be a section [T-3] as a final step where we switch from current TM side failure handling to CheckpointFailureManager. I understood [T1-5] more like an explanation of the motivation for [T2]. Maybe, we could merge T2/T3 into one step, but as Stefan says, it is strange to merge T1 and T3. I would stick to [T1-T2-T3] PR ordering.

Reply:
@StefanRRichter and @azagrebin thanks for your suggestions, they are valuable, I totally agree. I will change the design document and simplify this PR (T1).

Reply:
@StefanRRichter and @azagrebin based on your suggestion, I have reduced some changes and adjust the design document. For **T3** in the document, I have created a new issue FLINK-12209 to track it. What do you think about this PR now?

Reply:
@yanghua thanks, I will take another look!